flocker/node/agents/ebs.py

```diff
@@ -390,122 +410,221 @@ def ec2_client(region, zone, access_key_id, secret_access_key):
     """
-    def _run_with_logging(self, *args, **kwargs):
+    def _run_with_logging(*args, **kwargs):
         """
-        Run given boto.ec2.connection.EC2Connection method with exception
-        logging for ``EC2ResponseError``.
+        Run given boto3.ec2.ServiceResource method with exception
+        logging for ``ClientError``.
         """
-        original = getattr(self, original_name)
-        method = getattr(original, method_name)
+        def log_boto3_error(e):
+            BOTO_EC2RESPONSE_ERROR(
+                aws_code=e.response['Error']['Code'],
+                aws_message=unicode(e.response['Error']['Message']),
+                aws_request_id=e.response['ResponseMetadata']['RequestId'],
+            ).write()
+            raise
 
-        # Trace IBlockDeviceAPI ``method`` as Eliot Action.
-        # See https://clusterhq.atlassian.net/browse/FLOC-2054
-        # for ensuring all method arguments are serializable.
-        with AWS_ACTION(operation=[method_name, args, kwargs]):
+        with AWS_ACTION(operation=[method.__name__, args[1:], kwargs]):
             try:
                 return method(*args, **kwargs)
-            except EC2ResponseError as e:
-                BOTO_EC2RESPONSE_ERROR(
-                    aws_code=e.code,
-                    aws_message=e.message,
-                    aws_request_id=e.request_id,
-                ).write()
-                raise
+            except SSLError as e:
+                # We lost an idle connection. Try again; boto3 will open
+                # a new one.
+                try:
+                    return method(*args, **kwargs)
+                except ClientError as e:
+                    log_boto3_error(e)
+            except ClientError as e:
+                log_boto3_error(e)
     return _run_with_logging
 
 
-def boto_logger(*args, **kwargs):
-    """
-    Decorator to log all callable boto.ec2.connection.EC2Connection
-    methods.
-
-    :return: A function that will decorate all methods of the given
-        class with Boto exception logging.
-    """
-    def _class_decorator(cls):
-        for attr in EC2Connection.__dict__:
-            # Log wrap all callable methods except `__init__`.
-            if attr != '__init__':
-                attribute = getattr(EC2Connection, attr)
-                if callable(attribute):
-                    setattr(cls, attr,
-                            _boto_logged_method(attr, *args, **kwargs))
-        return cls
-    return _class_decorator
-
-
-@boto_logger("connection")
-class _LoggedBotoConnection(PClass):
-    """
-    Wrapper ``PClass`` around ``boto.ec2.connection.EC2Connection``
-    to facilitate logging of exceptions from Boto APIs.
-
-    :ivar boto.ec2.connection.EC2Connection connection: Object
-        representing connection to an EC2 instance with logged
-        ``EC2ConnectionError``.
-    """
-    connection = field(mandatory=True)
-
-
-class _EC2(PClass):
+class EC2Client(PClass):
```

Comment:
Is this class useful at all, as opposed to just having the EBS interactions directly within the IBlockDeviceAPI implementation? It's true that the logging tells you something... but it *doesn't* map directly to individual AWS interactions anyway so maybe not as much as we'd like. And "retry on SSL error" can happen half-way through a series of cloud interactions, so the whole *series* is restarted if an *individual* operation fails... which is somewhat unexpected and may lead to bugs.